src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+
+        // TODO:  This looks messy but it is fast. Research a cleaner way
+        // to do this which keeps high performance lookup.
+        //
+        // Fast lookup table to convert WINHTTP_AUTH constants to strings.
+        // WINHTTP_AUTH_SCHEME_BASIC = 0x00000001;
+        // WINHTTP_AUTH_SCHEME_DIGEST = 0x00000008;
+        // WINHTTP_AUTH_SCHEME_NEGOTIATE = 0x00000010;
+        private static readonly string[] s_authSchemeStringMapping =
+        {
+            null,
+            "Basic",
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            "Digest",
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            "Negotiate"
+        };
+
+        private static readonly uint[] s_authSchemePriorityOrder =
+        {
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_NEGOTIATE,
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_DIGEST,
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_BASIC
+        };
+
+        #region Constants
+        private const string ClientAuthenticationOID = "1.3.6.1.5.5.7.3.2";
+        private const string UriSchemeHttps = "https";
+        private const string HeaderNameContentLength = "Content-Length";
+        private const string HeaderNameContentEncoding = "Content-Encoding";
+        private const string HeaderNameCookie = "Cookie";
+        private const string HeaderNameSetCookie = "Set-Cookie";
+        private const string EncodingNameDeflate = "deflate";
+        private const string EncodingNameGzip = "gzip";
+        private const string EmptyCookieHeader = HeaderNameCookie + ":";
+        private static readonly string[] s_httpHeadersSeparator = { "\r\n" };
+        private static readonly TimeSpan s_maxTimeout = TimeSpan.FromMilliseconds(int.MaxValue);
+        private static readonly TimeSpan s_infiniteTimeout = Timeout.InfiniteTimeSpan;
+        #endregion
+
+        #region Fields
+        private object _lockObject = new object();
+        private bool _doManualDecompressionCheck = false;
+        private WinInetProxyHelper _proxyHelper = null;
+        private AutomaticRedirectionPolicy _automaticRedirectionPolicy = AutomaticRedirectionPolicy.DisallowHttpsToHttp;
+        private int _maxAutomaticRedirections = 50;
+        private DecompressionMethods _automaticDecompression = DecompressionMethods.Deflate | DecompressionMethods.GZip;
+        private CookieUsePolicy _cookieUsePolicy = CookieUsePolicy.UseInternalCookieStoreOnly;
+        private CookieContainer _cookieContainer = null;
+
+        // TODO: This current design uses a handler-wide lock to Add/Retrieve
+        // from the cache.  Need to improve this for next iteration in order
+        // to boost performance and scalability.
+        private CredentialCache _credentialCache = new CredentialCache();
+        private Object _credentialCacheLock = new object();
```

Comment:
A fine thing for a later pass: Object => object, String => string, etc. (https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md item 11)